sparse/file.go

```diff
@@ -33,195 +31,56 @@ func SetupFileIO(direct bool) {
+func WriteDataInterval(file *os.File, dataInterval Interval, data []byte) error {
+	log.Debug("writing data ...")
+	_, err := fileWriteAt(file, data, dataInterval.Begin)
 	if err != nil {
-		log.Fatal("Failed to open file for reading:", string(path), err)
+		errStr := fmt.Sprintf("Failed to write file using interval:%s, error: %s", dataInterval, err)
+		log.Error(errStr)
+		return fmt.Errorf(errStr)
 	}
-	defer file.Close()
-
-	for r := range fileStream {
-		switch r.Kind {
-		case SparseHole:
-			// Process hole
-			var hash, data []byte
-			unorderedStream <- HashedDataInterval{HashedInterval{r, hash}, data}
-
-		case SparseData:
-			// Read file data
-			data := make([]byte, r.Len())
-			status := true
-			n, err := fileReadAt(file, data, r.Begin)
-			if err != nil {
-				status = false
-				log.Error("File read error", status)
-			} else if int64(n) != r.Len() {
-				status = false
-				log.Error("File read underrun")
-			}
-			hasher := sha1.New()
-			hasher.Write(salt)
-			hasher.Write(data)
-			hash := hasher.Sum(nil)
-			unorderedStream <- HashedDataInterval{HashedInterval{r, hash}, data}
-		}
-	}
-	wgroup.Done() // indicate to other readers we are done
+	log.Debug("written data")
+	return nil
 }
 
-// FileWriter supports concurrent file reading
-// add this writer to wgroup before invoking
-func FileWriter(fileStream <-chan DataInterval, path string, wgroup *sync.WaitGroup) {
-	// open file
-	file, err := fileOpen(path, os.O_WRONLY, 0)
+func HashDataInterval(file *os.File, dataInterval Interval) ([]byte, error) {
+	data, err := ReadDataInterval(file, dataInterval)
 	if err != nil {
-		log.Fatal("Failed to open file for wroting:", string(path), err)
+		return nil, err
 	}
-	defer file.Close()
+	hasher := sha1.New()
+	hasher.Write(data)
+	hash := hasher.Sum(nil)
 
-	for r := range fileStream {
-		switch r.Kind {
-		case SparseHole:
-			log.Debug("trimming...")
-			err := PunchHole(file, r.Interval)
-			if err != nil {
-				log.Fatal("Failed to trim file")
-			}
-
-		case SparseData:
-			log.Debug("writing data...")
-			_, err = fileWriteAt(file, r.Data, r.Begin)
-			if err != nil {
-				log.Fatal("Failed to write file")
-			}
-		}
-	}
-	wgroup.Done()
-}
-
-// OrderIntervals puts back "out of order" read results
-func OrderIntervals(prefix string, unorderedStream <-chan HashedDataInterval, orderedStream chan<- HashedDataInterval) {
-	pos := int64(0)
-	m := make(map[int64]HashedDataInterval) // out of order completions
-	for r := range unorderedStream {
-		if pos == r.Begin {
-			// Handle "in order" range
-			log.Debug(prefix, r)
-			orderedStream <- r
-			pos = r.End
-		} else {
-			// push "out of order"" range
-			m[r.Begin] = r
-		}
-
-		// check the "out of order" stash for "in order"
-		for pop, existsNext := m[pos]; len(m) > 0 && existsNext; pop, existsNext = m[pos] {
-			// pop in order range
-			log.Debug(prefix, pop)
-			orderedStream <- pop
-			delete(m, pos)
-			pos = pop.End
-		}
-	}
-	close(orderedStream)
+	return hash, nil
```

Comment:
have you tried that ? That is shorter. But Go By Example uses the way here. https://gobyexample.com/sha1-hashes